### PR TITLE
Support the use of PHP 8.0 "mixed" type

### LINKED ISSUE #35579: [Issue] Support the use of PHP 8.0 "mixed" type
This issue is automatically created based on existing pull request: magento/magento2#35572: Support the use of PHP 8.0 "mixed" type

---------
Currently Magento does not compile when using the `mixed` type, neither as a parameter or return type. The generated Interceptor / Proxy / etc. classes try to make this union type nullable (`?mixed`). This will result in a fatal PHP error since the union type already includes `null`, and `null` is not nullable.

During compilation you will be encountered with the following error: `Type "mixed" cannot be nullable`

This patch checks if the type is `mixed` before adding the nullable (`?`) prefix.

### PR DESCRIPTION
**Pull request by pmzandbergen**
Currently Magento does not compile when using the `mixed` type, neither as a parameter or return type. The generated Interceptor / Proxy / etc. classes try to make this union type nullable (`?mixed`). This will result in a fatal PHP error since the union type already includes `null`, and `null` is not nullable.

During compilation you will be encountered with the following error: `Type "mixed" cannot be nullable`

This patch checks if the type is `mixed` before adding the nullable (`?`) prefix.

### Resolved issues:
1. [x] resolves magento/magento2#35579: Support the use of PHP 8.0 "mixed" type

### CODE REVIEW

**Comment by xmav** [comment]
Hi @pmzandbergen !
Could you please share your use case/code snipped to reproduce this issue ?

Thanks,
Maksym

**Comment by pmzandbergen** [comment]
> Hi @pmzandbergen ! Could you please share your use case/code snipped to reproduce this issue ?
> 
> Thanks, Maksym

Created a very simple example module @ https://github.com/pmzandbergen/some-module

Simply add it to a vanilla Magento 2.4.4 installation and run `bin/magento setup:di:compile`. Output:

```
% bin/magento setup:di:compile                        
Compilation was started.
Interceptors generation... 4/9 [============>---------------]  44% 4 secs 218.0 MiB
In Generator.php line 140:
                                                                                                                                                                                                                
  Type "mixed" cannot be nullable                                                                                                                                                                               
  Class Some\Module\Model\TestClass\Interceptor generation error: The requested class did not generate properly, because the 'generated' directory permission is read-only. If --- after running the 'bin/mage  
  nto setup:di:compile' CLI command when the 'generated' directory permission is set to write --- the requested class did not generate properly, then you must add the generated class object to the signature  
   of the related construct method, only.                                                                                                                                                                       
                                                                                                                                                                                                                

setup:di:compile
```

This is caused by Magento using `?mixed` as a parameter type for the Interceptor for `Some\Module\Model\TestClass`

**Comment by ihor-sviziev** [comment]
Hi @pmzandbergen,
Could you please cover your changes with integration tests?

**Comment by ihor-sviziev** [comment]
@magento run all tests
@magento create issue

**Comment by pmzandbergen** [comment]
> Hi @pmzandbergen,
> Could you please cover your changes with integration tests?

Of course. I'll add tests at the end of this week.

**Comment by pmzandbergen** [comment]
@ihor-sviziev sorry didn't had time to add tests yet, also holiday around the corner. I'll try to find some time afterwards.

**Comment by amenk** [comment]
We stumbled across the same error, and it seems to happen only in developer mode as soon as there is a class which uses `mixed` as a return type and has plugins.

**Comment by pmzandbergen** [comment]
@ihor-sviziev found some spare time, writing a test now :)

**Comment by pmzandbergen** [comment]
@ihor-sviziev added test based on the proxy generator (all other generators will work the same way).

Note: The test will be skipped on PHP version <= 8.0.0

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/ObjectManager/Test/Unit/Code/Generator/_files/SampleMixedProxy.txt [hunk 1]
```diff
@@ -0,0 +1,104 @@
+namespace Magento\Framework\ObjectManager\Code\Generator;
+
+/**
+ * Proxy class for @see \Magento\Framework\ObjectManager\Code\Generator\SampleMixed
+ */
+class SampleMixed_Proxy extends SampleMixed implements \Magento\Framework\ObjectManager\NoninterceptableInterface
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/ObjectManager/Test/Unit/Code/Generator/_files/SampleMixedProxy.txt:6]: I guess it should be
```suggest
class Proxy extends SampleMixed implements \Magento\Framework\ObjectManager\NoninterceptableInterface
```
So I expect to have a test failure now

Comment (pmzandbergen) [reply lib/internal/Magento/Framework/ObjectManager/Test/Unit/Code/Generator/_files/SampleMixedProxy.txt:6]: Test succeeded

Comment (ihor-sviziev) [reply lib/internal/Magento/Framework/ObjectManager/Test/Unit/Code/Generator/_files/SampleMixedProxy.txt:6]: Right, but it shouldn't due to the actual class should have a different class name, right?

Comment (pmzandbergen) [reply lib/internal/Magento/Framework/ObjectManager/Test/Unit/Code/Generator/_files/SampleMixedProxy.txt:6]: The test for the mixed type uses a new `Sample` class called `SampleMixed` (instead of `Sample`). Compare:
https://github.com/magento/magento2/blob/de131cb51103786b55fdaeb06046a93a8bd3b480/lib/internal/Magento/Framework/ObjectManager/Test/Unit/Code/Generator/_files/SampleMixedProxy.txt

With:
https://github.com/magento/magento2/blob/de131cb51103786b55fdaeb06046a93a8bd3b480/lib/internal/Magento/Framework/ObjectManager/Test/Unit/Code/Generator/_files/SampleProxy.txt

The later is from the original (existing) test, e.g.:
- `Sample` -> `Sample_Proxy` (existing test, not using the `mixed` type at all)
- `SampleMixed` -> `SampleMixed_Proxy` (new test using the `mixed` type)

**Review by ihor-sviziev — APPROVED** [review]

**Comment by ihor-sviziev** [comment]
@magento run Static Tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**
Install fresh vanilla Magento 2.4-develop

**Manual testing scenario:**

1. Go to https://github.com/pmzandbergen/some-module and download "some modules".
2. Copy the module to app/code (app/code/Some/Module/...)
3. Enable the module `bin/magento module:enable Some_Module`
4. Run compiler `bin/magento setup:di:compile`

**Before:** :heavy_multiplication_x: Getting error while compiling.
Screenshot from 2022-08-25 14-35-39

**After:** :heavy_check_mark:  No error while compilation.
Screenshot from 2022-08-25 14-45-31

**Additionally tested:**

1. Login as admin.
2. Create order from admin.
3. Login as customer.
4. Place an order from customer.
5. Created invoice,shipped and credit memo.
